Shivers Randomizer/App.xaml.cs

```diff
@@ -1489,51 +1493,82 @@ private async void ArchipelagoLoadData()
-            archipelago_Client?.SaveData("MetalDamage", ReadMemory(256, 1));
-            archipelago_Client?.SaveData("HealItemsReceived", archipelagoReceivedItems?.Count(num => num == (int)APItemID.FILLER.HEAL) ?? 0);
-
-            //Save Ixupi captures
-            int ixupiCaptured = ReadMemory(-60, 2);
-            archipelago_Client?.SaveData("IxupiCapturedStates", ReadMemory(-60, 2));
+            dataStorage.SkullDials["Prehistoric"] = dataStorage.SkullDials["Prehistoric"] with { Value = ReadMemory(836, 1) };
+            dataStorage.SkullDials["TarRiver"] = dataStorage.SkullDials["TarRiver"] with { Value = ReadMemory(840, 1) };
+            dataStorage.SkullDials["Werewolf"] = dataStorage.SkullDials["Werewolf"] with { Value = ReadMemory(844, 1) };
+            dataStorage.SkullDials["Burial"] = dataStorage.SkullDials["Burial"] with { Value = ReadMemory(848, 1) };
+            dataStorage.SkullDials["Egypt"] = dataStorage.SkullDials["Egypt"] with { Value = ReadMemory(852, 1) };
+            dataStorage.SkullDials["Gods"] = dataStorage.SkullDials["Gods"] with { Value = ReadMemory(856, 1) };
+
+            // Save Ixupi damage
+            dataStorage.IxupiDamage["Water"] = dataStorage.IxupiDamage["Water"] with { Value = ReadMemory(184, 1) };
+            dataStorage.IxupiDamage["Wax"] = dataStorage.IxupiDamage["Wax"] with { Value = ReadMemory(192, 1) };
+            dataStorage.IxupiDamage["Ash"] = dataStorage.IxupiDamage["Ash"] with { Value = ReadMemory(200, 1) };
+            dataStorage.IxupiDamage["Oil"] = dataStorage.IxupiDamage["Oil"] with { Value = ReadMemory(208, 1) };
+            dataStorage.IxupiDamage["Cloth"] = dataStorage.IxupiDamage["Cloth"] with { Value = ReadMemory(216, 1) };
+            dataStorage.IxupiDamage["Wood"] = dataStorage.IxupiDamage["Wood"] with { Value = ReadMemory(224, 1) };
+            dataStorage.IxupiDamage["Crystal"] = dataStorage.IxupiDamage["Crystal"] with { Value = ReadMemory(232, 1) };
+            dataStorage.IxupiDamage["Lightning"] = dataStorage.IxupiDamage["Lightning"] with { Value = ReadMemory(240, 1) };
+            dataStorage.IxupiDamage["Sand"] = dataStorage.IxupiDamage["Sand"] with { Value = ReadMemory(248, 1) };
+            dataStorage.IxupiDamage["Metal"] = dataStorage.IxupiDamage["Metal"] with { Value = ReadMemory(256, 1) };
+
+            dataStorage = dataStorage with
+            {
+                Jukebox = dataStorage.Jukebox || IsKthBitSet(ReadMemory(377, 1), 5),
+                TarRiverShortcut = dataStorage.TarRiverShortcut || IsKthBitSet(ReadMemory(368, 1), 6),
+                Health = ReadMemory(-40, 1),
+                HealItemsReceived = archipelagoReceivedItems?.Count(num => num == (int)APItemID.FILLER.HEAL) ?? 0,
+                IxupiCapturedStates = ReadMemory(-60, 2),
+                PuzzlesSolved = dataStorage.PuzzlesSolved with
+                {
+                    Gears = dataStorage.PuzzlesSolved.Gears || IsKthBitSet(ReadMemory(361, 1), 7),
```

Comment:
I should be able to improve a lot of this to reduce reads for those address locations that are the same byte.

Reply:
Also note the || on these is to prevent setting solves back to false.